File: jsoup/jsoupTester/src/main/java/com/html5libtester/jsoup/Html5libSerializer.java
package com.html5libtester.jsoup;

import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;

import org.jsoup.nodes.Attribute;
import org.jsoup.nodes.Comment;
import org.jsoup.nodes.DataNode;
import org.jsoup.nodes.Node;
import org.jsoup.nodes.TextNode;

public class Html5libSerializer {
	public static String dom2string(Node node) {
		String str = "";
		int ancestors = 0;
		if (node.childNode(0) == null)
			return "| ";
		Node parent = node;
		Node current = node.childNode(0);
		Node next = null;

		for (;;) {
			str += "\n| " + indent(ancestors + templateAncestors(current, 0));

			switch (current.getClass().toString()
					.replace("class org.jsoup.nodes.", "")) {
			case "Comment":
				Comment c = (Comment) current;
				str += "<!-- " + c.attr("comment") + " -->";
				break;
			case "DataNode":
				DataNode dn = (DataNode) current;
				str += '"' + dn.getWholeData() + '"';
				break;
			case "DocumentType":
				// DocumentType d = (DocumentType) current;
				String dname = current.attr("name");
				String publicid = current.attr("publicid");
				String systemid = current.attr("systemid");

				if (!publicid.isEmpty() && systemid.isEmpty()) {
					publicid = " \"" + publicid + "\"";
					systemid = " \"\"";
				} else if (publicid.isEmpty() && !systemid.isEmpty()) {
					publicid = " \"\"";
					systemid = " \"" + systemid + "\"";
				} else if (!publicid.isEmpty() && !systemid.isEmpty()) {
					publicid = " \"" + publicid + "\"";
					systemid = " \"" + systemid + "\"";
				}

				str += "<!DOCTYPE " + dname + publicid + systemid + '>';
				break;
			case "Element":
			case "FormElement":
				str += "<";

				// if (current.baseUri() != null)
				// switch (current.baseUri()) {
				// case "http://www.w3.org/2000/svg":
				// str += "svg ";
				// break;
				// case "http://www.w3.org/1998/Math/MathML":
				// str += "math ";
				// break;
				// }

				str += current.nodeName();
				str += '>';
				if (current.attributes().size() > 0) {
					Map<String, String> attrNames = new HashMap<String, String>();
					for (Attribute att : current.attributes().asList()) {
						String name = "";
						name += att.getKey();
						attrNames.put(name, att.getValue());
					}

					if (attrNames.size() > 0) {
						TreeMap<String, String> sorted_map = new TreeMap<String, String>(
								attrNames);
						for (Map.Entry<String, String> entry : sorted_map
								.entrySet()) {
							String key = entry.getKey();
							String value = entry.getValue();

							str += "\n| " + indent(1 + ancestors + templateAncestors(current, 0)) + key;
							str += "=\"" + value + "\"";
						}
					}
				}
				/*
				 * Template elements have 'content' child 
				 * This is the required format of html5lib
				 */
				if(current.nodeName().equals("template")){
					str += "\n| " + indent(1 + ancestors + templateAncestors(current, 0)) + "content";
				}
				break;
			case "Entities":
				str += "<![CDATA[ " + current.nodeName() + " ]]>";
				break;
			case "TextNode":
				TextNode t = (TextNode) current;
				str += '"' + t.getWholeText() + '"';
				break;

			default:
				break;
			}

			next = current.childNodeSize() > 0 ? current.childNode(0) : null;
			if (null != next) {
				parent = current;
				current = next;
				ancestors++;
				continue;
			}

			for (;;) {
				next = current.nextSibling();
				if (next != null) {
					current = next;
					break;
				}
				current = current.parent();
				parent = parent.parent();
				ancestors--;
				if (current == node) {
					return str.substring(1);
				}
			}
		}
	}

	private static String indent(int ancestors) {
		String str = "";
		if (ancestors > 0) {
			while (0 <= --ancestors)
				str += "  ";
		}
		return str;
	}

	private static int templateAncestors(Node current, int ancestors){
		if(current.parent() != null)
		{
			current = current.parent();
			if(current.nodeName().equals("template"))
				ancestors++;
			return templateAncestors(current, ancestors);
		}
		else
			return ancestors;
		
	}
}

File: AngleSharp/Html5libSerializer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AngleSharp.Dom;

namespace AngleSharpParser
{
    class Html5libSerializer
    {
        private readonly static String HTML = "http://www.w3.org/1999/xhtml";
        private readonly static String MATHML = "http://www.w3.org/1998/Math/MathML";
        private readonly static String SVG = "http://www.w3.org/2000/svg";
        private readonly static String XLink = "http://www.w3.org/1999/xlink";
        private readonly static String XML = "http://www.w3.org/XML/1998/namespace";
        private readonly static String XMLNS = "http://www.w3.org/2000/xmlns/";

        public static String dom2string(INode node)
        {
            String str = "";
            int ancestors = 0;
            if (!node.HasChildNodes)
                return "| ";
            INode parent = node;
            INode current = node.FirstChild;
            INode next = null;

            for (; ; )
            {
                str += "\n| " + indent(ancestors + templateAncestors(current, 0));

                switch (current.NodeType)
                {
                    case NodeType.Comment:
                        IComment c = (IComment)current;
                        str += "<!-- " + c.TextContent + " -->";
                        break;
                    case NodeType.DocumentType:
                        IDocumentType docType = (IDocumentType)current;
                        String dname = docType.Name;
                        String publicid = docType.PublicIdentifier;
                        String systemid = docType.SystemIdentifier;

                        if (!String.IsNullOrEmpty(publicid) && String.IsNullOrEmpty(systemid))
                        {
                            publicid = " \"" + publicid + "\"";
                            systemid = " \"\"";
                        }
                        else if (String.IsNullOrEmpty(publicid) && !String.IsNullOrEmpty(systemid))
                        {
                            publicid = " \"\"";
                            systemid = " \"" + systemid + "\"";
                        }
                        else if (!String.IsNullOrEmpty(publicid) && !String.IsNullOrEmpty(systemid))
                        {
                            publicid = " \"" + publicid + "\"";
                            systemid = " \"" + systemid + "\"";
                        }

                        str += "<!DOCTYPE " + dname + publicid + systemid + '>';
                        break;
                    case NodeType.Entity:
                        str += "<![CDATA[ " + current.NodeName + " ]]>";
                        break;
                    case NodeType.Text:
                        IText t = (IText)current;
                        str += '"' + t.Text + '"';
                        break;
                    case NodeType.Element:
                        IElement element = (IElement)current;
                        str += "<";

                        if (element.NamespaceUri != null)
                        {
                            if (element.NamespaceUri.Equals(SVG))
                                str += "svg ";
                            else if (element.NamespaceUri.Equals(MATHML))
                                str += "math ";
                        }

                        str += current.NodeName.ToLower();
                        str += '>';
                        if (element.Attributes.Count() > 0)
                        {
                            foreach (var att in element.Attributes.OrderBy(x => x.Name).ToList())
                            {
                                String key;
                                String value = att.Value;

                                if (att.NamespaceUri == null)
                                    key = att.Name;
                                else if (att.NamespaceUri.Equals(XML))
                                    key = "xml " + att.LocalName;
                                else if (att.NamespaceUri.Equals(XMLNS))
                                    key = "xmlns " + att.LocalName;
                                else if (att.NamespaceUri.Equals(XLink))
                                    key = "xlink " + att.LocalName;
                                else
                                    key = att.Name;


                                str += "\n| " + indent(1 + ancestors + templateAncestors(element, 0)) + key;
                                str += "=\"" + value + "\"";
                            }
                        }
                        /*
                         * Template elements have 'content' child 
                         * This is the required format of html5lib
                         */
                        if (element.NodeName.Equals("template") && element.NamespaceUri != null && element.NamespaceUri.Equals(HTML))
                        {
                            str += "\n| " + indent(1 + ancestors + templateAncestors(current, 0)) + "content";
                        }
                        break;


                    default:
                        break;
                }

                next = current.FirstChild;
                if (null != next)
                {
                    parent = current;
                    current = next;
                    ancestors++;
                    continue;
                }

                for (; ; )
                {
                    next = current.NextSibling;
                    if (next != null)
                    {
                        current = next;
                        break;
                    }
                    current = current.Parent;
                    parent = parent.Parent;
                    ancestors--;
                    if (current == node)
                    {
                        return str.Substring(1);
                    }
                }
            }
        }

        private static String indent(int ancestors)
        {
            String str = "";
            if (ancestors > 0)
            {
                while (0 <= --ancestors)
                    str += "  ";
            }
            return str;
        }

        private static int templateAncestors(INode current, int ancestors)
        {
            if (current.Parent != null)
            {
                current = current.Parent;
                if (current.NodeName.Equals("template") && ((IElement)current).NamespaceUri != null
                    && ((IElement)current).NamespaceUri.Equals(HTML))
                    ancestors++;
                return templateAncestors(current, ancestors);
            }
            else
                return ancestors;

        }
    }
}
